File: NotificationAPI.Infrastructure/Data/Repositories/UnitOfWork.cs
using NotificationAPI.Domain.Repositories;

namespace NotificationAPI.Infrastructure.Data.Repositories;

public class UnitOfWork : IUnitOfWork
{
    private readonly ApplicationDbContext _context;
    private IUserRepository _users;
    private INotificationRepository _notifications;

    public UnitOfWork(ApplicationDbContext context)
    {
        _context = context;
    }

    public IUserRepository Users => _users ??= new UserRepository(_context);
    public INotificationRepository Notifications => _notifications ??= new NotificationRepository(_context);

    public async Task<int> SaveChangesAsync()
    {
        return await _context.SaveChangesAsync();
    }

    public void Dispose()
    {
        _context?.Dispose();
    }
}

File: NotificationAPI.Infrastructure/Data/Repositories/UnitOfWork.vb
Imports NotificationAPI.Domain.Repositories

Namespace NotificationAPI.Infrastructure.Data.Repositories

    Public Class UnitOfWork
        Implements IUnitOfWork

        Private _context As ApplicationDbContext
        Private _users As IUserRepository
        Private _notifications As INotificationRepository

        Public Sub New(context As ApplicationDbContext)
            _context = context
        End Sub

        Public ReadOnly Property Users As IUserRepository Implements IUnitOfWork.Users
            Get
                If _users Is Nothing Then
                    _users = New UserRepository(_context)
                End If
                Return _users
            End Get
        End Property

        Public ReadOnly Property Notifications As INotificationRepository Implements IUnitOfWork.Notifications
            Get
                If _notifications Is Nothing Then
                    _notifications = New NotificationRepository(_context)
                End If
                Return _notifications
            End Get
        End Property

        Public Async Function SaveChangesAsync() As Task(Of Integer) Implements IUnitOfWork.SaveChangesAsync
            Return Await _context.SaveChangesAsync()
        End Function

        Public Sub Dispose() Implements IUnitOfWork.Dispose
            _context?.Dispose()
        End Sub
    End Class

End Namespace
